selfdrive/boardd/tests/test_boardd_usbprotocol.cc

```diff
@@ -0,0 +1,127 @@
+    std::string bytes(data_len, '\0');
+    std::generate(bytes.begin(), bytes.end(), std::ref(rbe));
+    test_data[data_len] = bytes;
+  }
+
+  // generate can messages for this panda
+  int other_panda_message_cnt = 3;
+  auto can_list = msg.initEvent().initSendcan(can_list_size + other_panda_message_cnt);
+  for (uint8_t i = 0; i < can_list_size; ++i) {
+    auto can = can_list[i];
+    uint32_t id = random_int(0, std::size(dlc_to_len) - 1);
+    const std::string &dat = test_data[dlc_to_len[id]];
+    can.setAddress(i);
+    can.setSrc(bus_offset);
+    can.setDat(kj::ArrayPtr((uint8_t *)dat.data(), dat.size()));
+    total_pakets_size += CANPACKET_HEAD_SIZE + dat.size();
+  }
+
+  // generate can messages from other panda
+  for (int i = can_list_size; i < can_list_size + other_panda_message_cnt; ++i) {
+    auto can = can_list[i];
+    uint32_t id = random_int(0, std::size(dlc_to_len) - 1);
+    const std::string &dat = test_data[dlc_to_len[id]];
+    can.setAddress(i);
+    can.setSrc(bus_offset + PANDA_BUS_CNT);
+    can.setDat(kj::ArrayPtr((uint8_t *)dat.data(), dat.size()));
+  }
+
+  can_data_list = can_list.asReader();
+  INFO("test " << can_list_size << " packets, total size " << total_pakets_size);
+}
+
+void PandaTest::test_can_send() {
+  std::vector<uint8_t> unpacked_data;
+  this->pack_can_buffer(can_data_list, [&](uint8_t *chunk, size_t size) {
+    int size_left = size;
+    for (int i = 0, counter = 0; i < size; i += CHUNK_SIZE, counter++) {
+      REQUIRE(chunk[i] == counter);
+
+      const int len = std::min(CHUNK_SIZE, size_left);
+      unpacked_data.insert(unpacked_data.end(), &chunk[i + 1], &chunk[i + len]);
+      size_left -= len;
+    }
+  });
+  REQUIRE(unpacked_data.size() == total_pakets_size);
+
+  int cnt = 0;
+  INFO("test can message integrity");
+  for (int pos = 0, pckt_len = 0; pos < unpacked_data.size(); pos += pckt_len) {
+    can_header header;
+    memcpy(&header, &unpacked_data[pos], CANPACKET_HEAD_SIZE);
+    const uint8_t data_len = dlc_to_len[header.data_len_code];
+    pckt_len = CANPACKET_HEAD_SIZE + data_len;
+
+    REQUIRE(header.addr == cnt);
+    REQUIRE(test_data.find(data_len) != test_data.end());
+    const std::string &dat = test_data[data_len];
+    REQUIRE(memcmp(dat.data(), &unpacked_data[pos + 5], dat.size()) == 0);
+    ++cnt;
+  }
+  REQUIRE(cnt == can_list_size);
+}
+
+void PandaTest::test_can_recv() {
+  std::vector<can_frame> frames;
+  this->pack_can_buffer(can_data_list, [&](uint8_t *data, size_t size) {
+    this->unpack_can_buffer(data, size, frames);
+  });
+
+  REQUIRE(frames.size() == can_list_size);
+  for (int i = 0; i < frames.size(); ++i) {
+    REQUIRE(frames[i].address == i);
+    REQUIRE(test_data.find(frames[i].dat.size()) != test_data.end());
+    const std::string &dat = test_data[frames[i].dat.size()];
+    REQUIRE(memcmp(dat.data(), frames[i].dat.data(), dat.size()) == 0);
+  }
+}
+
+TEST_CASE("send/recv can packets") {
+  auto bus_offset = GENERATE(2, 4);
```

Comment:
why is it between 2 and 4?

Reply:
I set it arbitrarily, what value do you think is more reasonable

Reply:
bus_offset = index * PANDA_BUS_CNT in our code, than offset is being added to(for can_receive) or substracted from(for can_send) bus.
so for first panda offset = 0 and for second = 4.